### ISSUE TITLE
File Reload of natlink should restart the whole python "subsystem"

### ISSUE THREAD

**Issue by quintijn** [dictation-toolbox/natlink#28]

**Comment by quintijn** [comment]
At line 200, the reloadPython is performed, I think this is the function that is called when the File, Reload button is pressed in the Messages window.

It would be nice, when the whole python "subsystem" is closed down and restarted, so all underlying modules are also refreshed (reloaded) after changes. Hope this is possible to realize at some time.

**Comment by LexiconCode** [comment]
If this were to be implemented please read the following issue https://github.com/dictation-toolbox/natlink/pull/129#issuecomment-1235433191 as it has implications for restarting the Python subsystem.

**Comment by drmfinlay** [comment]
This is what the Reload button is supposed to do, right?  I could look into fixing it.

**Comment by LexiconCode** [comment]
> This is what the Reload button is supposed to do, right?  I could look into fixing it.

That's my understanding, it's meant to reload the the interpreter.

**Comment by drmfinlay** [comment]
All right then.  Looks like a job for me.

**Comment by drmfinlay** [comment]
Okay, I've looked into this now.

The reload button is broken because it only reloads the 'natlinkcore' module.  This newer module is not a replacement for the old 'natlinkmain' module, which the button used to reload, so it doesn't work right.  In Natlink 5, pressing reload should result in the re-running of the 'natlinkcore' loader.  I'll submit a pull request to address this issue.

**Comment by drmfinlay** [comment]
Yes, that should be possible.  That is what the button used to do, at least theoretically.  According to the log, it was changed to only reload 'natlinkmain' in October, 2019.  (See commit d05715c.)  It looks like this was a workaround.

**Comment by drmfinlay** [comment]
I managed to get the button to reload the interpreter.  However, PyWin32 does not support this sort of reinitialisation.  Its COM component raises errors that prevent most of my command modules from loading.  This is an old issue, apparently, which probably won't be fixed.

Since most of us rely on PyWin32, the best Natlink's reload button can do reload Natlink's Python code and any user modules.

**Comment by drmfinlay** [comment]
My proposal from Wednesday to use Natlink exclusively in normal Python processes (#198), if realised, has implications for the reload button.  The simplest way to deal with it, I think, is to add a new callback to be used in 'natlinkcore':

    def onReload():
        # Reload user modules.
        # ...

    natlink.setReloadCallback(onReload)

**Comment by drmfinlay** [comment]
The issue I rediscovered above with interpreter reinitialization has implications for Natlink's Python thread support.  If the reload button cannot practically reinitialize the interpreter, thread support might just as well be enabled by default.

**Comment by quintijn** [comment]
Why not do so? Honestly, I never saw any difference in the working of Natlink with or without the threading option. If you think on is OK, just put the default on!!

**Comment by drmfinlay** [comment]
All right then, that can be changed in a future release.  I would suggest major version 6 include that and my implementation of #198, if it works as expected.

It looks like thread support was only ever optional because of the interpreter reinitialization.  The effect of it being disabled is noticeable only if another thread is started from Python.
